ui/ui.go

```diff
@@ -96,27 +96,39 @@ func (ui *WriterUI) PrintTable(table Table) {
 	}
 }
 
-func (ui *WriterUI) AskForText(label string) (string, error) {
-	var text string
-
-	err := interact.NewInteraction(label).Resolve(&text)
-	if err != nil {
-		return "", fmt.Errorf("Asking for text: %s", err)
+func (ui *WriterUI) AskForText(opts TextOpts) (string, error) {
+	if opts.ValidateFunc == nil {
+		opts.ValidateFunc = func(s string) (bool, string, error) {
+			return true, "", nil
+		}
 	}
 
-	return text, nil
+	for {
+		text := opts.Default
+		err := interact.NewInteraction(opts.Label).Resolve(&text)
+		if err != nil {
+			return "", fmt.Errorf("Asking for text: %s", err)
+		}
+
+		isValid, message, err := opts.ValidateFunc(text)
+		if err != nil {
+			return "", fmt.Errorf("Validation error: %s", message)
+		}
+		if isValid {
+			return text, nil
+		}
```

Comment:
this should be a bit different actually. if there is an error, it's not really to do with validation result but rather with the process of validation itself, so should return error, but since we are asking user to retry we should also tell them what went wrong.

```
if err != nil {
	return "", fmt.Errorf("Validation input: %s", err)
}
if isValid {
	return text, nil
} else {
        if len(message) == 0 {
            message = "(reason for failure not specified)"
        }
        ui.ErrorLinef("Failed validation: %s", message)
}
```

Reply:
Sounds Good. Adopted.